File: BFS/BFS.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <bitset>
using namespace std;

bitset<100001> vis(0);
queue<int> bfs;

#define vvi vector<vector<int>>

    


void BFS(vvi &graph, int start){

    
    cout<<start<<" ";
    bfs.push(start); vis.set(start);

    while(!bfs.empty()){

        int curr = bfs.front();
        bfs.pop();

        for(int neigh: graph[curr]){

            if(!vis.test(neigh)){

                cout<<neigh<<" ";
                bfs.push(neigh); vis.set(neigh);

            }
        }

    }cout<<'\n';

}


int main(){

    int A, C, K, x, y;
    cin>>A>>C>>K;
    vector<vector<int>> graph(A, vector<int>(0));
    

    for(int i = 0; i<C; i++){
        
        cin>>x>>y;
        graph[x].push_back(y);
        graph[y].push_back(x);

    }

    BFS(graph, K);
    return 0;
}

File: BFS/BFS.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class BFS {

    // Declaration of global variables
    static BitSet vis = new BitSet(100001);
    static Queue<Integer> bfs = new ArrayDeque<>();

    // Definition of a data type to represent a graph as an adjacency list
    static class Graph {

        Vector<Vector<Integer>> adjList;

        Graph(int size) {
            adjList = new Vector<>(size);
            for (int i = 0; i < size; i++) {
                adjList.add(new Vector<>());
            }
        }

        void addEdge(int src, int dest) {
            adjList.get(src).add(dest);
            // adjList.get(dest).add(src);  //in case of an undirected graph
        }
    }



  

    // Implementation of the breadth-first search algorithm (BFS)
    static void BFS(Graph graph, int start) {

        System.out.print(start + " ");
        bfs.add(start);
        vis.set(start);

        while (!bfs.isEmpty()) {
            int curr = bfs.poll();

            for (int neigh : graph.adjList.get(curr)) {
                if (!vis.get(neigh)) {  // if neigh is not visited
                    System.out.print(neigh + " ");
                    bfs.add(neigh);    // add it to the queue
                    vis.set(neigh);    // mark as visited
                }
            }
        }
        System.out.println();
    }





  
    public static void main(String[] args) throws IOException {

        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));

        int N, M, Start, x, y;

        // Input data
        StringTokenizer tokenizer = new StringTokenizer(reader.readLine());
        N = Integer.parseInt(tokenizer.nextToken());  // Number of vertices
        M = Integer.parseInt(tokenizer.nextToken());  // Number of edges
        Start = Integer.parseInt(tokenizer.nextToken());  // Start vertex

        // Create the graph
        Graph graph = new Graph(N);

        // Input the edges and build the graph
        for (int i = 0; i < M; i++) {
            tokenizer = new StringTokenizer(reader.readLine());
            x = Integer.parseInt(tokenizer.nextToken());
            y = Integer.parseInt(tokenizer.nextToken());
            graph.addEdge(x, y);
        }

        // Call the BFS function
        BFS(graph, Start);
    }
}




